### ISSUE TITLE
lr_scheduler doesn't work when start training from checkpoint

### ISSUE THREAD

**Issue by stvreumi** [NervanaSystems/distiller#70]
Hello:
I want to use `MultiStepMultiGammaLR` scheduler in my pruning lr_scheduler.
When I using the `compress_classifier.py` to pruning the `res_net20_cifar` from the begining and define the lr_scheduler in the `yaml` file, it works well.
But when I using the checkpoint to train and prune, the lr_scheduler defined in the `yaml` file doesn't work. The lr **doesn't decay** when the epoch achieve defined milestone.

I use the script below:
`python3 compress_classifier.py --arch resnet20_cifar  dataset/ -p=50 --lr=0.3 --epochs=150 -b 128 --compress=resnet20_cifar_ele_pruning.yaml -j=1 --vs 0 --deterministic --resume=logs/resnet20_cifar_baseline/checkpoint.pth.tar`

Below is my `yaml` setting
```
version: 1
pruners:
  low_pruner:
    class: AutomatedGradualPruner
    initial_sparsity : 0.05
    final_sparsity: 0.60
    weights: [module.layer1.2.conv1.weight,  module.layer1.2.conv1.weight,
              module.layer1.0.conv1.weight,  module.layer1.0.conv2.weight,
              module.layer1.1.conv1.weight,  module.layer1.1.conv2.weight]

  mid_pruner:
    class:  AutomatedGradualPruner
    initial_sparsity : 0.05
    final_sparsity: 0.67
    weights: [module.layer2.2.conv1.weight,  module.layer2.2.conv2.weight,
              module.layer2.0.conv2.weight,  module.layer2.0.downsample.1.weight,
              module.layer2.0.conv1.weight,  module.layer2.0.downsample.0.weight,
              module.layer2.1.conv1.weight,  module.layer2.1.conv2.weight]

  high_pruner:
    class:  AutomatedGradualPruner
    initial_sparsity : 0.05
    final_sparsity: 0.76
    weights: [module.layer3.0.conv1.weight,  module.layer3.1.conv1.weight,
              module.layer3.1.conv2.weight,  module.layer3.0.conv2.weight,
              module.layer3.0.downsample.0.weight, module.layer3.0.downsample.1.weight,
              module.fc.weight]
lr_schedulers:
  training_lr:
    class: MultiStepMultiGammaLR
    milestones: [300, 302, 400]
    gammas: [0.1, 0.1, 0.5]

policies:
    - pruner:
        instance_name: low_pruner
      starting_epoch: 300
      ending_epoch: 400
      frequency: 2
    - pruner:
        instance_name: mid_pruner
      starting_epoch: 300
      ending_epoch: 400
      frequency: 2
    - pruner:
        instance_name: high_pruner
      starting_epoch: 300
      ending_epoch: 400
      frequency: 2
    - lr_scheduler:
        instance_name: training_lr
      starting_epoch: 0
      ending_epoch: 400
      frequency: 1
```
Is there any problem in my script and `yaml` setting?

**Comment by hunterkun** [comment]
I guess it is because you  ignore the start epoch is not 0 if you resume model from checkpoint file instead it is the last epoch you train last time. so the lr scheduler may work on epoch 300+start epoch, 302+.., 400+..]. i hope this will solve your problem.

**Comment by nzmora** [comment]
Hi @stvreumi  and @hunterkun,

Kun's answer was also what I initially thought, but the answer is a bit more complicated.
If you look at PyTorch's [LRScheduler implementation](https://pytorch.org/docs/stable/_modules/torch/optim/lr_scheduler.html), you will see that the basic mechanism is pretty simple: each time ```step()``` is invoked they increase a counter (```self.last_epoch```) and then update the LR.  So ```self.last_epoch``` tell the LR scheduler what is the current epoch.

Now we get back to Kun's answer: you are resuming a checkpoint at epoch=180, but when ```MultiStepMultiGammaLR``` is instantiated, the constructor argument ```last_epoch``` is not explicitly set and so it gets the value ```-1``` because this is the default value ([see code](https://github.com/NervanaSystems/distiller/blob/c247b57f6f14e609b39cb3e8c1fbae29ca0fe049/distiller/learning_rate.py#L52)).
So ```MultiStepMultiGammaLR``` starts counting from 0, not from 180.

The fastest and simplest way around this is to tell ```MultiStepMultiGammaLR```  that we are starting at epoch=180:
```
lr_schedulers:
  training_lr:
    class: MultiStepMultiGammaLR
    milestones: [300, 302, 400]
    gammas: [0.1, 0.1, 0.5]
    last_epoch: 180   <============================ I added this line
```

This is annoying, and error prone, so I think the better solution is to determine what is the epoch we are starting from, and to pass this to the constructor of the LR policy.
The relevant code is [here](https://github.com/NervanaSystems/distiller/blob/c247b57f6f14e609b39cb3e8c1fbae29ca0fe049/distiller/config.py#L97) and [here](https://github.com/NervanaSystems/distiller/blob/c247b57f6f14e609b39cb3e8c1fbae29ca0fe049/distiller/config.py#L112).  
If either one of you wants to take a stab at this and issue a PR with the fix, I would be happy ;-).

Cheers!
Neta

**Comment by hunterkun** [comment]
@nzmora I try to implement your idea that is to add new  parameter named `start_epoch` to `file_config` function and then the calling 
`lr_schedulers = __factory('lr_schedulers', model, sched_dict, optimizer=optimizer, last_epoch=start_epoch)`
but i encounter a problem: if we explicitly set last_epoch value other than -1, then pytorch will think you want to resume a optimizer which has 'initial_lr' attribute. However, our optimizer obviously don't have this. so it would give KeyError.
```
if last_epoch == -1:
     for group in optimizer.param_groups:
           group.setdefault('initial_lr', group['lr'])
else:
      for i, group in enumerate(optimizer.param_groups):
             if 'initial_lr' not in group:
                  raise KeyError("param 'initial_lr' is not specified "
                           "in param_groups[{}] when resuming an optimizer".format(i))
```

**Comment by nzmora** [comment]
@hunterkun I haven't tested this yet, but I think we need both (1) + (2), or (1) + my initial suggestion (setting the epoch in LRScheduler constructor).

If we do (2) alone, then the LR of the first epoch (until we take the first ```step()```) will be wrong.
This code in the constructor of _LRScheduler takes a ```step()```:
https://github.com/pytorch/pytorch/blob/3749c581b79cba49f511b19fa02c0f50fa05b250/torch/optim/lr_scheduler.py#L22

And in the ```step()``` function the _LRScheduler sets the Optimzer's initial LR:
https://github.com/pytorch/pytorch/blob/3749c581b79cba49f511b19fa02c0f50fa05b250/torch/optim/lr_scheduler.py#L51-L56

I'm also thinking of a different solution: (2) + invoke ```step(<first epoch from checkpoint>)``` after constructing the _LRScheduler.  I'm not sure I like this solution.  It is simple but somewhat brittle.

**Comment by hunterkun** [comment]
@nzmora,  you think the initial lr is still what we set using command line rather than that of last training time we saved when we resume, isn't it? If so, this small change indeed consider it. And it is sensible that we save and load the optimizer and then it is possible that we can resume training from the last lr in the last time.

**Comment by guyjacob** [comment]
Considering the points raised here and in @hunterkun's PR (#72), I think it would be best to just go ahead and implement loading of the optimizer state. Note that we're already saving it today.
(In fact, some optimizers (not SGD) actually maintain and update an internal state other than the LR. So if we're looking to properly support resumed training, we should be loading it anyway).

Following @nzmora's comment about the LR in the first epoch after resuming, it seems to me the solution has to be (1) + passing the epoch number in the scheduler's CTOR. (1) + (2) won't work. Then we could also "play it safe" and pass the epoch number in each call to step as in (2), but I don't really see a point in doing that.

**Comment by guyjacob** [comment]
@hunterkun (or @stvreumi) would you like to tackle this?

**Comment by stvreumi** [comment]
First I want to thank @hunterkun and @nzmora, currently the script is working as I expected after changing  the milestone epoch.

And, I took some time to research the comment above...
As far as I understand from @guyjacob, the next thing we need to do is
 **(1) + passing the epoch number in the scheduler's CTOR, right?**
And I have a problem is that, if we resume the optimizer from the checkpoint, what happening when the optimizer is conflict with the lr_scheduler?

I'm new to this repo. If I misunderstand anything, please let me know, thanks.
And I'm in the midterm week, sorry for replying late.

**Comment by guyjacob** [comment]
@stvreumi no worries about late replies, thanks for taking the time to contribute :)

I'm not sure what you mean by optimizer conflicting with the LR scheduler. The optimizer only stores the actual learning rate value. The LR scheduler takes the optimizer in its CTOR and initializes the learning rate according to the optimizer. Where do you see a possible conflict?

**Comment by stvreumi** [comment]
@guyjacob I have this idea because I see a lot of data in the optimizer value(ex. momentum buffer) when loading the checkpoint rather than just learning rate. So I have a question that, if I use different scheduler or learning rate from the checkpoint, will this cause any conflict?

**Comment by stvreumi** [comment]
OK I see, I will try my best to figure out how to fix the problem when I have time.

**Comment by levzlotnik** [comment]
Hi,
Closing this due to staleness. Feel free to reopen if needed.
